### PR TITLE
Update ncurses to 6.1, keep v5 ABI

### PR DESCRIPTION
**Pull request by Mno-hime**
Tested various ncurses programs (`nano`, `joe`, `tig`, `mc`, `lynx`).

Kept only some terminfo definitions (list got from https://src.fedoraproject.org/rpms/ncurses/blob/master/f/ncurses.spec#_183). Tested MATE Terminal and Xterm.

### CODE REVIEW

**Comment by pyhalov** [comment]
This one will likely break ncurses application on console - see #3579

**Comment by Mno-hime** [comment]
~Yes, ncurses frames are broken. Fixed with https://github.com/tsoome/illumos-gate/commits/gfx. We may be able to push this soonish.~

**Comment by Mno-hime** [comment]
Added changes from https://github.com/OpenIndiana/oi-userland/pull/3579, pulled patches from OmniOS CE, updated to 6.1.20190209.

**Comment by Mno-hime** [comment]
Updated to 6.1.20190518: https://invisible-island.net/ncurses/NEWS.html#t20190518.

**Comment by Mno-hime** [comment]
Anything's missing here?

**Comment by pyhalov** [comment]
Does it work fine on console (graphic, text, regression on com console)?

**Comment by Mno-hime** [comment]
> Does it work fine on console (graphic, text

Yes, I run several ncurses programs in console and Xterm.

> regression on com console)?

Good idea, something I'll have to a look at.

**Comment by Mno-hime** [comment]
> > regression on com console)?
> 
> Good idea, something I'll have to a look at.

So, I exported VirtualBox guest console on COM1 via TCP, connected by `nc` and with both 5.9 and 6.1 stuff is not that usable: can't scroll `man` nor `mc`, weird things happening allover the place...

Do you know how to setup a more sane test virtual environment? I don't have HW serial console anywhere.

**Comment by pyhalov** [comment]
What if you set terminal to something simple like vt100? Scrolling can be broken on serial port. Are there any differences between 5.9 and 6.1? What about window borders? Do they look sane?

**Comment by Mno-hime** [comment]
Scrolling works there, frames look OK, but numbers 2..10 of `Fx` shortcuts in `mc` are not visible for 6.1:

**ncurses 5.9 console sun-color**
mc-nc59-console
**ncurses 6.1.20190518 console sun-color**
mc-nc61-console
**ncurses 6.1.20190518 serial console xterm-color256**
mc-nc61-serial

I'll report it upstream.

Exporting `TERM=vt100` is not changing the thing for serial; the echo of what I type is not helpful as well:
```
newman@server:~$ echo $TERM
echo $TERM
vt100
```

**Comment by pyhalov** [comment]
Any news on this?

**Comment by Mno-hime** [comment]
> Any news on this?

Extending supported colors range at the first hunk of https://github.com/OpenIndiana/oi-userland/pull/4745/files#diff-4ff81593a708379cd198497b430f0d06R11 breaks the F2-F10 keys color of MC (the second image). 

Without the hunk it's fine.

Talked with the ncurses maintainer Thomas E. Dickey about this. And this is what he has to say about it:

```
yes - I can see a problem using unmap from

https://invisible-island.net/misc_tools/index.html#item:unmap

for example

\E[40m 4
\E[30m
\E[46mEdit    
\E[315m
\E[40m 5
\E[30m
\E[46mCopy    
\E[315m
\E[40m 6
\E[30m
\E[46mRenMov 
\E[315m
\E[40m 7
\E[30m
\E[46mMkdir   
\E[315m
\E[40m 8

The "m" codes are SGR (set graphic rendition), which is used to set colors.
The "315" is not a legal value (unless someone's been changing the console
emulator).  Legal values would be 30 to 39 for the foreground (text) color.
The background colors are 40 to 49.

Since they're not legal values, the console terminal emulator simply
ignores them, leaving the last _legal_ value as "30" (black).

Black-on-black isn't very readable.
```

Any ideas @tsoome ?

**Comment by tsoome** [comment]
So why are they generating the \E[315m sequences in first place?

**Comment by Mno-hime** [comment]
https://github.com/OpenIndiana/oi-userland/pull/5054 is a workaround for the `mc` problem, once merged this PR can be merged as well. Also updated to June 1st version.

**Comment by pyhalov** [comment]
If extending terminal definition breaks terminal color interpretation, I think it shouldn't be applied. 
I don't understand what generates '\E[315m' , but if ncurses is wrong, why should we fix mc instead of fixing ncurses?

**Comment by Mno-hime** [comment]
I removed the 256 color support part, rebuilt and re-tested MC. Please, review.

Also forwarded @tsoome conversation with the ncurses maintainer. Perhaps he infers some clues from that. I can't tell anything sensible about the `\E[315m` problem.

File: components/library/ncurses/patches/02-xterm-no-rep.patch [hunk 1]
```diff
@@ -0,0 +1,20 @@
+Although a true xterm has supported the ECMA-48 REP (repeat character) control
+since January 1997, many other terminal emulators that claim to be xterm do
+not support this. (For example, VTE, konsole, PuTTY, iTerm2).
+
+For least surprise, remove this capability from
+the xterm terminal definition in OmniOS.
```
Comment (pyhalov) [inline components/library/ncurses/patches/02-xterm-no-rep.patch]: s/in OmniOS//

**Comment by Mno-hime** [comment]
@pyhalov @tsoome here's the comparison of what we ship in illumos and in ncurses 6.1:
```diff
--- sun-color.illumos   2019-06-07 20:54:27.767277356 +0000
+++ sun-color.ncurses   2019-06-07 20:54:29.574889366 +0000
@@ -1,18 +1,22 @@
 sun-color|Sun Microsystems Workstation console with color support (IA systems)
-colors#256
+colors#8
 ncv#3
-pairs#32767
+pairs#64
+bold=\E[1m
 cub=\E[%p1%dD
 cud=\E[%p1%dB
 cuf=\E[%p1%dC
 cuu=\E[%p1%dA
 home=\E[H
 op=\E[0m
+rs2=\E[s
 setab=\E[4%p1%dm
 setaf=\E[3%p1%dm
 setb=\E[4%?%p1%{1}%=%t4%e%p1%{3}%=%t6%e%p1%{4}%=%t1%e%p1%{6}%=%t3%e%p1%d%;m
 setf=\E[3%?%p1%{1}%=%t4%e%p1%{3}%=%t6%e%p1%{4}%=%t1%e%p1%{6}%=%t3%e%p1%d%;m
-smso=\E[1m
+sgr=\E[0%?%p6%t;1%;%?%p1%p3%|%t;7%;m
+sgr0=\E[m
+smso=\E[7m
 smul=\E[4m
 rmul=\E[24m
 use=sun
```

**Comment by tsoome** [comment]
> @pyhalov @tsoome here's the comparison of what we ship in illumos and in ncurses 6.1:
> 
> ```diff
> --- sun-color.illumos   2019-06-07 20:54:27.767277356 +0000
> +++ sun-color.ncurses   2019-06-07 20:54:29.574889366 +0000
> @@ -1,18 +1,22 @@
>  sun-color|Sun Microsystems Workstation console with color support (IA systems)
> -colors#256
> +colors#8
>  ncv#3
> -pairs#32767
> +pairs#64
> +bold=\E[1m
>  cub=\E[%p1%dD
>  cud=\E[%p1%dB
>  cuf=\E[%p1%dC
>  cuu=\E[%p1%dA
>  home=\E[H
>  op=\E[0m
> +rs2=\E[s
>  setab=\E[4%p1%dm
>  setaf=\E[3%p1%dm

This is why we got \e [ 315 m, the mc did emit color 15 (light white) as p1. 

>  setb=\E[4%?%p1%{1}%=%t4%e%p1%{3}%=%t6%e%p1%{4}%=%t1%e%p1%{6}%=%t3%e%p1%d%;m
>  setf=\E[3%?%p1%{1}%=%t4%e%p1%{3}%=%t6%e%p1%{4}%=%t1%e%p1%{6}%=%t3%e%p1%d%;m
> -smso=\E[1m
> +sgr=\E[0%?%p6%t;1%;%?%p1%p3%|%t;7%;m
> +sgr0=\E[m
> +smso=\E[7m
>  smul=\E[4m
>  rmul=\E[24m
>  use=sun
> ```

The tem and sample terminfo patches are posted for review, those will enable \e [ 38 ; 5 (and 48) sequence, which will allow to specify colors 0 - 255 directly, so the colors 0-15 will be fine for all modes. 15-255 are only available for depth 24 and 32.

**Comment by Mno-hime** [comment]
I updated `01-sun-color.patch` with changes from https://www.illumos.org/rb/r/1983/diff/1/ and `mc` is now happy in 256 colors and otherwise in 32-bit framebuffer console.

**Comment by Mno-hime** [comment]
@tsoome Should we wait with this for "11213 tem: partially support CSI [ 38 and 48 sequences with smaller depths"?

**Comment by tsoome** [comment]
> @tsoome Should we wait with this for "11213 tem: partially support CSI [ 38 and 48 sequences with smaller depths"?

Yes, or pick the entries from xterm-256, as those are implementing color checks in terminfo language and are using 30-37, 40-47 and corresponding bright sequences.

**Comment by pyhalov** [comment]
What does block this?

**Comment by Mno-hime** [comment]
> What does block this?

Someone should look at the terminfo set I selected for publication (I picked those I saw in Fedora and openSUSE ncurses packages and what I myself felt is important [Microsoft terminals, vscode, all Xterm, alacrytty, ...]) and see if something valuable isn't missing.

**Comment by pyhalov** [comment]
Do you mean 01-sun-color.patch?

**Comment by Mno-hime** [comment]
I mean what terminfo files are to be shipped, I picked only some in interrest of space: https://github.com/OpenIndiana/oi-userland/pull/4745/files#diff-7808e589c67c8f5cfb6491029116f825L1026.

**Comment by normjacobs** [comment]
FWIW, you might consider packaging the "base" terminfo files in one package and anything extra in another for folks that might install if support for the ancient terminal that they just rescued from the trash heap.  Just a thought and not necessarily this PR

**Comment by pyhalov** [comment]
@Mno-hime If you mean that ncurses ships too much terminfo data, move everything you consider not-so-useful to some library/ncurses/terminfo-extra package or include it all with ncurses as it is done now.

**Comment by Mno-hime** [comment]
Will make my life easier and ship it everything, not worth the 1.5 MB saved. Tested again `mc` and `mc -S modarin256`.

I think it's ready for merge.

**Pull request merged**